fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1625,6 +1679,187 @@ Future<Void> configureTenant(Reference<DB> db,
+				return Void();
+			} else {
+				CODE_PROBE(true, "Metacluster rename new name already exists");
+				throw tenant_already_exists();
+			};
+		} else {
+			if (self->tenantId == -1) {
+				self->tenantId = oldTenantEntry.get().id;
+			}
+			wait(self->ctx.setCluster(tr, oldTenantEntry.get().assignedCluster.get()));
+			if (oldTenantEntry.get().tenantState != TenantState::READY) {
+				CODE_PROBE(true, "Metacluster unable to proceed with rename operation");
+				throw invalid_tenant_state();
+			}
+		}
+
+		if (oldTenantEntry.get().id != self->tenantId) {
+			// The tenant must have been removed simultaneously
+			CODE_PROBE(true, "Metacluster rename old tenant ID mismatch");
+			throw tenant_removed();
+		}
+
+		// If marked for deletion, abort the rename
+		// newTenantEntry.get().tenantState == TenantState::REMOVING is already checked above
+		if (oldTenantEntry.get().tenantState == TenantState::REMOVING) {
+			CODE_PROBE(true, "Metacluster rename candidates marked for deletion");
+			throw tenant_removed();
+		}
+
+		TenantMapEntry updatedOldEntry = oldTenantEntry.get();
+		TenantMapEntry updatedNewEntry(updatedOldEntry);
+		updatedOldEntry.tenantState = TenantState::RENAMING_FROM;
+		updatedNewEntry.tenantState = TenantState::RENAMING_TO;
+		updatedOldEntry.renamePair = self->newName;
+		updatedNewEntry.renamePair = self->oldName;
+
+		ManagementClusterMetadata::tenantMetadata.tenantMap.set(tr, self->oldName, updatedOldEntry);
+		ManagementClusterMetadata::tenantMetadata.tenantMap.set(tr, self->newName, updatedNewEntry);
+		return Void();
+	}
+
+	ACTOR static Future<Void> finishRenameFromManagementCluster(RenameTenantImpl* self,
+	                                                            Reference<typename DB::TransactionT> tr) {
+		state Optional<TenantMapEntry> oldTenantEntry;
+		state Optional<TenantMapEntry> newTenantEntry;
+		wait(store(oldTenantEntry, tryGetTenantTransaction(tr, self->oldName)) &&
+		     store(newTenantEntry, tryGetTenantTransaction(tr, self->newName)));
+
+		// Another (or several other) operations have already removed/changed the old entry
+		// Possible for the new entry to also have been tampered with,
+		// so it may or may not be present with or without the same id, which are all
+		// legal states. Assume the rename completed properly in this case
+		if (!oldTenantEntry.present() || oldTenantEntry.get().id != self->tenantId) {
+			CODE_PROBE(true, "Metacluster finished rename with missing entries or mismatched id");
+			return Void();
+		}
+		if (oldTenantEntry.get().tenantState == TenantState::REMOVING) {
+			throw tenant_removed();
+		}
+		ASSERT(newTenantEntry.present());
+		ASSERT(newTenantEntry.get().id == self->tenantId);
+
+		TenantMapEntry updatedOldEntry = oldTenantEntry.get();
+		TenantMapEntry updatedNewEntry = newTenantEntry.get();
+
+		// If marked as removing, we will not update the new entry
+		if (updatedNewEntry.tenantState != TenantState::REMOVING) {
```

Comment:
This should only update the entry if it is in the `RENAMING_TO` state. Otherwise, it's possible that the rename already finished (moving it to `READY`) and then switched to something else (e.g. `UPDATING_CONFIGURATION`).

I suppose we have an even more odd possible sequence, which is that we could:

1. Rename tenant1 -> tenant2
2. Rename tenant2-> tenant1
3. Rename tenant1 -> tenant2

If one client attempting the first rename stalls for a while, it may try to complete it after step 2 and be allowed to do so even if the data cluster has not been updated.

The easiest way around this would be to make use of the `configurationSequenceNum` field to prevent making updates if the tenant has already advanced beyond our current attempted modification.